### PR TITLE
 Configure Pyright

### PR DESCRIPTION
**Pull request by victorlin**
## Description of proposed changes

Add the necessary configuration including rule exceptions to make the linter run without error on the code as-is. This means the check is mostly useless except to prevent any violations of rules that aren't currently violated, but it allows for incremental adoption of rules rather than needing to address all violations at once.

List of exceptions determined using command:

    npx pyright --outputjson \
    | jq --raw-output '.generalDiagnostics.[] | .rule' \
    | sort -u

Pyright supports some checks and type inference that Mypy does not¹.

¹ https://github.com/microsoft/pyright/blob/1.1.316/docs/mypy-comparison.md

## Checklist

- [x] Pyright is run via Pytest in CI ([evidence](https://github.com/nextstrain/augur/actions/runs/9134727760/job/25120859369?pr=1246#step:8:742))
- [x] Checks pass
- [x] ~Add a message in [CHANGES.md](https://github.com/nextstrain/augur/blob/HEAD/CHANGES.md) summarizing the changes in this PR that are end user focused. Keep headers and formatting consistent with the rest of the file.~ no functional changes
- [ ] Post-merge: create issue to audit rule exceptions individually

### CODE REVIEW

**Comment by jameshadfield** [comment]
(1 year later - I must have missed this at the time)

I think this is a great direction to go in. My not-extensive understanding of type checking in python indicates pyright is preferred over mypy. It seems strange to me to use both typecheckers in the same project, but this seems to be rather common in python.

File: pyrightconfig.json
Comment (victorlin) [inline pyrightconfig.json:1]: I've changed the direction of this PR from trying to address all violations to just setting up Pyright and ignoring violations explicitly. This reflects similar work done in https://github.com/nextstrain/nextstrain.org/pull/778.

Here's commits from the old direction: 9e8c9e08cba0fde1d4c06b55bb31b3552ed5c82c...346983123edddaa0889aaf0e27a0c993a94fa23e and new direction: fda57c54086b8716197ecaccc767f381025d3023

File: docs/contribute/DEV_DOCS.md [hunk 1]
```diff
@@ -149,7 +149,15 @@ During development you can run static type checks using [mypy][]:
...
-There are also many [editor integrations for mypy][].
+and [pyright][]:
+
+    $ npx pyright
+    ...
+    Found 40 source files
+    0 errors, 0 warnings, 0 infos
+    Completed in 2sec
+
+There are also many [editor integrations for mypy][], and Pyright is integrated into VS Code's Python support.
```
Comment (jameshadfield) [inline docs/contribute/DEV_DOCS.md:160]: FWIW, to get this working nicely in VS Code I needed:
* Using [pylance](https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-pylance) (I'm guessing most people will have this already)
* Setting `python.analysis.typeCheckingMode = basic` (default: off)

I think the general view of the group has been to not document any editor-specific stuff, but I'd support a sentence or two about how to get it working in vs-code.

File: tests/test_pyright.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import pytest
+from pathlib import Path
+from shutil import which
+from subprocess import run
+
+topdir = Path(__file__).resolve().parent.parent
+
+pyright = which("pyright")
```
Comment (jameshadfield) [inline tests/test_pyright.py:8]: How about adding this as a dev dependency in `setup.py` (as we do for `mypy` and `flake8`), which would then make `pyright` available to all developers, and avoid the if/else + test skipping added here. We'd also avoid using `npm/npx` within augur, which feels strange to me. (This also applies to the changes in `DEV_DOCS.md`.)

File: pyrightconfig.json [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  "pythonVersion": "3.8",
+  "include": ["augur/"],
+
+
+  // These are rules that are enabled by default and currently have violations.
+
+  // TODO: go through these exceptions and determine if they should be kept or
+  // removed and violations addressed.
+
+  // TODO: consider aligning this with the same file in nextstrain/cli:
```
Comment (jameshadfield) [inline pyrightconfig.json:11]: I looked at a few of these - totally fine if we want to defer any and all to separate PRs beyond this one:
* `reportUnusedExpression` - single failure, should fix
* `reportUndefinedVariable` - single failure, already with a flake8 ignore comment
* `reportPrivateImportUsage` - no failures
* `reportGeneralTypeIssues` - 9 errors, but all seem worth fixing

File: tests/test_pyright.py (hunk 2 above)
Comment (genehack) [reply tests/test_pyright.py:8]: +1 to not crossing the streams

**Review by genehack — COMMENTED** [review]
Looks good; concur with James that we should add pyright as an explicit dev-dep.

File: tests/test_pyright.py (hunk 2 above)
Comment (tsibley) [reply tests/test_pyright.py:8]: The `pyright` distribution on PyPI is unofficial and all it does is set up Node and invoke npm behind the scenes to install the official NPM `pyright` distribution. It's still "crossing the streams", just behind your back.

It seems more straightforward and less prone to issues with Node envs to do this ourselves instead. I personally don't really see an issue with mixing package ecosystems/managers when they're disjoint.

(Background: Victor got this approach from my setup for Nextstrain CLI. The `pyright` package on PyPI was introduced right around the same time I started using Pyright in the CLI, and I'm not sure if I was aware of then or not.)

Comment (victorlin) [reply tests/test_pyright.py:8]: Continuing in #1468

**Review by genehack — APPROVED** [review]
good to go, yeah?

File: pyrightconfig.json (hunk 3 above)
Comment (victorlin) [reply pyrightconfig.json:11]: #1472

**Pull request merged**